rolling-shutter/shdb/notify.go

```diff
@@ -0,0 +1,79 @@
+package shdb
+
+import (
+	"context"
+
+	"github.com/jackc/pgconn"
+	"github.com/jackc/pgx/v4"
+	"github.com/rs/zerolog/log"
+)
+
+// SlurpNotifications waits for notifications from the postgres database and puts them onto the
+// given channel in a loop.
+func SlurpNotifications(ctx context.Context, conn *pgx.Conn, chann chan<- *pgconn.Notification) {
+	for {
+		notification, err := conn.WaitForNotification(ctx)
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			if err != nil {
+				log.Error().Err(err).Msg("error waiting for notification")
+				continue
+			}
+			log.Info().Str("channel", notification.Channel).Msg("database notification received")
+			chann <- notification
+		}
+	}
+}
+
+func ExecListenChannels(ctx context.Context, conn *pgx.Conn, channels []string) error {
+	for _, dbch := range channels {
+		_, err := conn.Exec(ctx, "listen "+dbch)
+		if err != nil {
+			log.Error().
+				Err(err).
+				Str("channel", dbch).
+				Msg("error listening to channel")
+			return err
+		}
+	}
+	return nil
+}
+
+type (
+	SignalLoopFunc func() error
+	SignalFunc     func()
+	SignalHandler  func(ctx context.Context) error
+)
+
+// NewSignal creates a new 'Signal'. This can be used to asynchronously call the given signal
+// handler from another go routine. A signal can be delivered while the signal handler is running,
+// but we will not run multiple signal handler calls in parallel. Instead the signal handler will
+// be called once for all signals delivered, while the handler was active.
+func NewSignal(ctx context.Context, signalName string, handler SignalHandler) (SignalFunc, SignalLoopFunc) {
+	logger := log.With().Str("signal", signalName).Logger()
+	ch := make(chan struct{}, 1)
+	loop := func() error {
+		for {
+			select {
+			case <-ch:
+				err := handler(ctx)
+				if err != nil {
+					logger.Info().Err(err).Msg("error calling handler function")
+				}
+			case <-ctx.Done():
+				return ctx.Err()
+			}
+		}
+	}
+
+	signal := func() {
+		select {
+		case ch <- struct{}{}:
```

Comment:
This code was meant to handle the notifications from the database.
We use a single database connection to receive those notifications.
When dispatching notifications I don't want to block the go routine reading the notifications. That's why the signal function should be non-blocking. It probably wouldn't matter much if we only ever had one type of database notification, but I don't want the processing of one type of notification block the other types.

Reply:
> If the channel in the loop is blocked (handler blocked and an additional signal in the channel), this will drop into the default case and thus the signal is dropped forever. Thus we could miss insert events in the database or notifications from other handlers.
> 
> I'm not sure there is a case in the business-logic where one particular handler can have 2 unprocessed signals in their pipeline, but if that would happen there might be the chance that the cycle of signals might be broken without being able to recover. But I haven't figured out a condition where this would actually happen.

The channel we create does have a buffer size of 1, so we should be able to send a signal while the handler is active.
And that's all we need and want here. The handler function is supposed to work on all relevant changes  that happened since it has been called the last time. So, it's fine to call the handler only once even though we may have called the signal function multiple times.

Reply:
You are right, making this blocking would be worse.

> The channel we create does have a buffer size of 1, so we should be able to send a signal while the handler is active.


Yes, I tried to figure out a scenario where the buffer size of 1 would not be sufficient to keep up with the signal back-pressure, leading to dropped signals. 
If the dropped signal is *not* the last signal to have called the handler function that's fine, since the overall flow of signals is not broken forever and can resume. And as you say the handler functions can deal with the relevant changes.
This is e.g. the case in the `newDecryptionKey` handler, since the handler is not only called on database inserts but is also manually called after the submit-batch handler is finished.

So I was worried about a case where a dropped signal would starve the liveness of the overall chain of signals. 

The send-decryption-trigger handler is only dependent on the database insert signal. But here I think the back-pressure on the channel will never be high enough to drop a signal, since the batcher's batch-closing loop is highly synchronised on the sequencer's batch number, which in turn is dependent on the overall batch-submission.

I think we're good to go to merge it as is.

If we discover an edge-case where the signal-chain would be broken, we could always introduce additional keep-alive calls to the handler functions.